sched/adaptdl_sched/allocator.py

```diff
@@ -38,11 +38,90 @@ class AdaptDLAllocator(object):
+                    if event["type"] == "ADDED":
+                        await self.lock.acquire()
+                        # re-read the job , compared with the previous readjob
+                        job_old = event["object"]
+                        namespace_old = job_old["metadata"]["namespace"]
+                        name_old = job_old["metadata"]["name"]
+
+                        job_list = await \
+                            self._objs_api.list_namespaced_custom_object(
+                                "adaptdl.petuum.com", "v1", "", "adaptdljobs")
+
+                        job = None
+                        for job_new in job_list["items"]:
+                            namespace_new = job_new["metadata"]["namespace"]
+                            name_new = job_new["metadata"]["name"]
+                            # found the old job
+                            if namespace_new == namespace_old \
+                                    and name_new == name_old:
+                                job = job_new
+                                break
+                        else:
+                            # This job is don
+                            self.lock.release()
+                            continue
+
+                        # some other coroutine has handled this
+                        LOG.info(job.get("status", {}).get("allocation", []))
+                        if len(job.get("status", {})
+                                .get("allocation", [])) != 0:
+                            # release the lock before continuing
+                            self.lock.release()
+                            continue
+
+                        namespace = job["metadata"]["namespace"]
+                        name = job["metadata"]["name"]
+                        # if this is a restarted job, skip i
+                        LOG.info("detected an added job %s/%s.",
+                                 namespace, name)
+                        # parse the job infomation
+                        job_info = {}
+                        self._get_job_info(job, job_info)
+
+                        # find available nodes
+                        node_info_list, _ = await self._find_nodes()
+
+                        # get the node to allocate
+                        new_allocation = self._policy._allocate_job(
+                                            job_info, node_info_list)
+                        # allocate the job on the suggest_node
+                        if new_allocation:
+                            patch = {"status": {"allocation": new_allocation}}
+                            LOG.info("Patch AdaptdlJob %s/%s: %s ",
+                                     namespace, name, patch)
+                            await patch_job_status(self._objs_api, namespace,
+                                                   name, patch)
+                        # release lock
+                        self.lock.release()
```

Comment:
Use `async with` rather than `acquire` and `release`

Reply:
Also suggest splitting out the inner contents of the loop into a different method.